semantics/src/main/kotlin/com/strumenta/kolasu/semantics/indexing/IndexReader.kt

```diff
@@ -0,0 +1,75 @@
+package com.strumenta.kolasu.semantics.indexing
+
+import kotlin.reflect.KClass
+import kotlin.reflect.safeCast
+
+/**
+ * Component handling read operations over an index.
```

Comment:
It is not clear why we need this? Can't we just read directly from the index? Is this a sort of cache?

Reply:
The index reader provides a read-only view over an index. This component is available when defining a reference resolver, type computer, scope computer, etc. The objective is to prevent writes and allows type-safe queries. 

The idea is that we can have an specific index - the actual storage depending on the implementation - and generic readers and writers providing advanced r/w features regardless of the underlying technology.

Does it make sense?

Reply:
To me this seems similar to the `Collection` interface (the immutable variant). I understand the "offering a subset of Index" part, but as Alessio I would think of it as the base interface from which Index would be derived. Alternatively it could be a specialization of Index which just throw exceptions when the writing methods are called. Similarly to Index seems to me to bee over-generic. It also does not seem to have any method specific to indexes. If anything it seems to me a "ReadOnlyCollection". Perhaps you need it because you need to use the PublishedAPI annotation. If that is the case I would explain that in the documentation. Maybe one could present it as a sort of ReadOnlyCollection needed exclusively to support the DSL API and not intended to be reused for other goals

Reply:
I am not sure I understand why this looks similar to the _Collection_ interface. 
The current code is a proposal, feel free to change it as you wish.

Reply:
Well, a Collection has additional requirements I think, at least in Java, so it makes sense that one doesn't have to implement all Collection methods just to have access to the index/storage. That said, I too have the feeling that this design is a little too abstract. Maybe we could do without index readers and writers, and have the index present a writable view and a read-only view? I.e. `Index` by default is writable, but if you want to protect it from writes, you call a `.readOnlyView()` method on it (or access a `.readOnlyView` getter) that returns an instance wrapping the index and throwing exceptions on all methods that would mutate it. Kind of what happens with `Collection.unmodifiableCollection()` in Java. That's just a suggestion.

Reply:
After discussion we decided to call this `RepositoryReader`